File: src/main/java/kwartz/Element.java
/**
 *  @(#) Element.java
 *  @Id  $Id$
 *  @copyright $Copyright$
 *  @release $Release$
 */
package kwartz;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.Iterator;

public class Element {
    private String _name;
    private Tag    _stag;
    private Tag    _etag;
    private List   _cont;           // list of Statement
    private PresentationDeclaration _decl;
    private BlockStatement _plogic;

    public Element(String name, Tag stag, Tag etag, List cont) {
        _name = name;
        _stag = stag;
        _etag = etag;
        _cont = cont;
    }

    public Tag getStag() { return _stag; }
    public Tag getEtag() { return _etag; }
    public List getCont() { return _cont; }
    public String getName() { return _name; }

    public static Map createElementTable(List elementList) {
        return addElementList(new HashMap(), elementList);
    }

    public static Map addElementList(Map elementTable, List elementList) {
        for (Iterator it = elementList.iterator(); it.hasNext(); ) {
            Element elem = (Element)it.next();
            elementTable.put(elem.getName(), elem);
        }
        return elementTable;
    }

    public static void mergeDeclarationList(Map elementTable, List declList) {
        // merge declarations
        if (elementTable == null) return;
        if (declList == null) return;
        for (Iterator it = declList.iterator(); it.hasNext(); ) {
            PresentationDeclaration decl = (PresentationDeclaration)it.next();
            for (int i = 0; i < decl.names.length; i++) {
                String name = decl.names[i];
                Element elem = (Element)elementTable.get(name);
                if (elem != null)  elem.setDeclaration(decl);
            }
        }
    }



    public PresentationDeclaration getDeclaration() {
        return _decl;
    }

    public void setDeclaration(PresentationDeclaration decl) {
        _decl = decl;
        if (decl.value != null) {
            Expression[] args = { decl.value };
            PrintStatement stmt = new PrintStatement(args);
            _cont = new ArrayList();
            _cont.add(stmt);
        }
        if (decl.tagname != null) {
            // TBI
            if (decl.tagname.getToken() == TokenType.STRING) {
                String tagname = ((StringExpression)decl.tagname).getValue();
                _stag.tagname = tagname;
                _etag.tagname = tagname;
            }
        }
        if (decl.remove != null) {
            if (_stag.attrs != null) {
                for (int i = _stag.attrs.size() - 1; i >= 0; i--) {
                    Attr attr = (Attr)_stag.attrs.get(i);
                    Object aname = attr.name;
                    if (decl.remove.contains(aname))
                       _stag.attrs.remove(i);
                }
            }
        }
        if (decl.attrs != null) {
            if (_stag.attrs == null) _stag.attrs = new ArrayList();
            for (Iterator it = decl.attrs.keySet().iterator(); it.hasNext(); ) {
                String aname = (String)it.next();
                Object expr  = decl.attrs.get(aname);
                int i;
                int len = _stag.attrs.size();
                for (i = 0; i < len; i++) {
                    Attr attr = (Attr)_stag.attrs.get(i);
                    if (aname.equals(attr.name)) {
                        attr.value = expr;
                        break;
                    }
                }
                if (i >= len) {
                    Attr attr = new Attr(" ", aname, expr);
                    _stag.attrs.add(attr);
                }
            }
            //for (int i = _stag.attrs.size() -1; i >= 0; i--) {
            //    Attr attr = (Attr)_stag.attrs.get(i);
            //    String aname = attr.name;
            //    Object expr = decl.attrs.get(aname);
            //    if (expr != null)  attr.value = expr;
            //}
        }
        if (decl.append != null) {
            if (_stag.append_exprs == null)
                _stag.append_exprs = decl.append;
            else
                _stag.append_exprs.addAll(decl.append);
        }
        if (decl.plogic != null) {
            _plogic = decl.plogic;
        }
    }

    public BlockStatement getPresentationLogic() {
        if (_plogic == null) {
            Statement[] stmts = {
                new ExpandStatement(ExpandStatement.STAG),
                new ExpandStatement(ExpandStatement.CONT),
                new ExpandStatement(ExpandStatement.ETAG),
            };
            _plogic = new BlockStatement(stmts);
        }
        return _plogic;
    }


    public Statement[] getContentStatements() {
        Statement[] stmts = new Statement[_cont.size()];
        _cont.toArray(stmts);
        return stmts;
    }


    public StringBuffer _inspect() {
        return _inspect(0, new StringBuffer());
    }

    public StringBuffer _inspect(int level, StringBuffer sb) {
        for (int i = 0; i < level; i++) sb.append("  ");
        sb.append("Element(");
        sb.append(_name);
        sb.append(")\n");
        for (Iterator it = _cont.iterator(); it.hasNext(); ) {
            Statement stmt = (Statement)it.next();
            stmt._inspect(level + 1, sb);
        }
        return sb;
    }
}

File: lib/kwartz/element.rb
###
### copyright(c) 2005 kuwata-lab all rights reserved
###
### $Id$
###

module Kwartz

   class Element
      def initialize(name, tagname, content_stmt, attrs={}, append_expr=[], is_empty=false, spaces=['','','',''])
         @name     = name
         @tagname  = tagname
         @content  = content_stmt
         @attrs    = attrs       || {}
         @append   = append_expr || []
         @is_empty = is_empty
         @spaces   = spaces
         list = [ ExpandStatement.new(:stag), ExpandStatement.new(:cont), ExpandStatement.new(:etag) ]
         @plogic   = BlockStatement.new(list)
         @remove   = []
      end
      attr_accessor :name, :tagname, :content, :attrs, :append, :is_empty, :spaces, :plogic
      alias :marking :name

      def swallow(elem_decl)
         return unless @name == elem_decl.name
         @tagname = elem_decl.tagname if elem_decl.tagname
         @content = PrintStatement.new([elem_decl.value]) if elem_decl.value
         elem_decl.attrs.each do |key,val|
            @attrs[key] = val
         end if elem_decl.attrs
         @append.concat(elem_decl.append) if elem_decl.append
         @remove.concat(elem_decl.remove) if elem_decl.remove
         @plogic = elem_decl.plogic if elem_decl.plogic
      end
      
      def self.create_from_taginfo(name, staginfo, etaginfo, body_stmt_list)
         Kwartz::assert unless !etaginfo || staginfo[:tagname] == etaginfo[:tagname]
         tagname      = staginfo[:tagname]
         content      = BlockStatement.new(body_stmt_list)
         attrs        = staginfo[:attr_values]
         append_expr  = staginfo[:append_expr]
         is_empty     = staginfo[:is_empty]
         spaces       = [  staginfo[:before_space],
                           staginfo[:after_space],
                           etaginfo ? etaginfo[:before_space] : '',
                           etaginfo ? etaginfo[:after_space]  : '',
                        ]
         stag_begend  = [ staginfo[:is_begline], staginfo[:is_endline ], ]
         etag_begend  = etaginfo ? [ etaginfo[:is_begline], etaginfo[:is_endline ], ] : nil
         return Element.new(name, tagname, content, attrs, append_expr, is_empty, spaces)
      end
      
      def stag_stmt(properties={})
         arguments = []
         if @tagname.is_a?(String)
            arguments << StringExpression.new("#{@spaces[0]}<#{@tagname}")
         elsif @tagname.is_a?(Expression)
            arguments << StringExpression.new("#{@spaces[0]}<")
            arguments << @tagname
         else
            Kwartz::assert
         end
         @attrs.each do |aname, avalue|
            if @remove && @remove.include?(aname)
               next
            elsif avalue.is_a?(String)
               arguments << StringExpression.new(" #{aname}=\"#{avalue}\"")
            elsif avalue.is_a?(Expression)
               arguments << StringExpression.new(" #{aname}=\"")
               arguments << avalue
               arguments << StringExpression.new("\"")
            else
               Kwartz::assert
            end
         end
         @append.each do |expr|
            arguments << expr
         end if @append
         s = (@is_empty && properties[:html] != true) ? ' /' : ''
         arguments << StringExpression.new("#{s}>#{spaces[1]}")
         return PrintStatement.new(arguments)
      end
      
      def cont_stmt(properties={})
         #Kwartz::assert unless @is_empty && @content
         return @content
      end
         
      def etag_stmt(properties={})
         #return nil if @is_empty
         return BlockStatement.new([]) if @is_empty   # for bug 1110250
         arguments = []
         if @tagname.is_a?(String)
            arguments << StringExpression.new("#{@spaces[2]}</#{@tagname}>#{@spaces[3]}")
         elsif @tagname.is_a?(Expression)
            arguments << StringExpression.new("#{@spaces[2]}</")
            arguments << @tagname
            arguments << StringExpression.new(">#{@space[3]}")
         else
            Kwartz::assert
         end
         return PrintStatement.new(arguments)
      end

      
      def statement(properties={})
         list = []
         list << stag_stmt(properties)
         stmt = cont_stmt(properties)
         if stmt.token == :block
            list.concat(stmt.statements)
         else
            list << stmt
         end
         list << etag_stmt(properties)
         return BlockStatement.new(list)
      end
      
      def _inspect()
         s = ''
         s << "===== marking=#{@name} =====\n"
         s << "[tagname]\n#{@tagname}#{@is_empty ? '/':''}\n"
         s << "[attrs]\n"
         @attrs.each do |key,val|
            s << "#{key}=" << (val.is_a?(Expression) ? val._inspect : '"'+val.to_s+"\"\n")
         end
         if @append && !@append.empty?
            s << "[append]\n"
            @append.each do |expr|
               s << expr._inspect
            end
         end
         if @remove && !@remove.empty?
            s << "[remove]\n"
            @remove.each_with_index do |str, i|
               s << " " if i > 0
               s << str.dump
            end
            s << "\n"
         end
         s << "[content]\n" << @content._inspect       if @content
         s << "[spaces]\n"  << @spaces.inspect << "\n" if @spaces
         s << "[plogic]\n"  << @plogic._inspect        if @plogic
         return s
      end

      ## returns element_table (hash of element.name => element)
      def self.merge(element_list, element_decl_list)
         decl_table = {}
         element_decl_list.each do |decl|
            decl_table[decl.name] = decl
         end if element_decl_list
         element_table = {}
         element_list.each do |elem|
            decl = decl_table[elem.name]
            elem.swallow(decl) if decl
            element_table[elem.name] = elem
         end if element_list
         return element_table
      end
      
   end

   class ElementDeclaration
      def initialize(name, value=nil, attrs={}, append=[], remove=[], tagname=nil, plogic=nil)
         @name    = name
         @value   = value
         @attrs   = attrs
         @append  = append
         @remove  = remove
         @tagname = tagname
         @plogic  = plogic
      end

      attr_reader :name, :value, :attrs, :append, :remove, :tagname, :plogic
      alias :marking :name
   
      def self.create_from_hash(name, hash)
         name    = name
         value   = hash[:value]
         attrs   = hash[:attr] || hash[:attrs] || {}
         append  = hash[:append] || []
         remove  = hash[:remove] || []
         tagname = hash[:tagname]
         plogic  = hash[:plogic]
         return self.new(name, value, attrs, append, remove, tagname, plogic)
      end

      def _inspect()
         s = "\##{@name} {\n"
         s <<    "  value:\n" << @value._inspect(2) if @value
         if @attrs && !@attrs.empty?
            s << "  attrs:\n"
            @attrs.keys.sort.each do |key|
               s << "    \"#{key}\" " << @attrs[key]._inspect()
            end
         end
         #s << "  attrs:\n    "  << @attrs.keys.sort.collect { |key| "\"#{key}\" " + @attrs[key]._inspect() }.join("    ") if @attrs && !@attrs.empty?
         if @append
            s << "  append:\n"
            @append.each do |expr|
               s << expr._inspect(2)
            end
         end
         s << "  remove:\n    " << @remove.sort.collect{|aname| '"'+aname+'"'}.join(",") << "\n" if @remove && !@remove.empty?
         s << "  tagname:\n" << @tagname._inspect(2) if @tagname
         s << "  plogic:\n" << @plogic._inspect(2)  if @plogic
         s << "}\n"
         return s
      end
      
   end

end
